paddle/framework/recurrent_network_op.cc

```diff
@@ -0,0 +1,224 @@
+  PADDLE_ENFORCE(memories.size() == boot_memories.size(),
+                 "the size of memories and boot_memories doesn't match: %d,%d",
+                 memories.size(), boot_memories.size());
+  for (size_t i = 0; i < memories.size(); ++i) {
+    MemoryAttr mem_attr;
+    mem_attr.var = memories[i];
+    mem_attr.pre_var = pre_memories[i];
+    mem_attr.boot_var = boot_memories[i];
+    memory_attrs_.push_back(mem_attr);
+    LOG(INFO) << "set memorys:\t"
+              << "memory:" << mem_attr.var << "\tboot:" << mem_attr.boot_var;
+  }
+}
+
+void RecurrentOp::CreateScopes(ScopePtr scope) const {
+  LOG(INFO) << "create scopes";
+  auto dims = Input(scope, 0)->GetMutable<Tensor>()->dims();
+  size_t seq_len = dims[0];
+  Variable* scopes_var = scope->GetVariable(step_scopes_name_);
+  auto step_scopes = scopes_var->GetMutable<std::vector<ScopePtr>>();
+  // TODO Only two scopes are needed for inference, this case will be
+  // supported later.
+  if (seq_len > step_scopes->size()) {
+    for (size_t i = step_scopes->size(); i < seq_len; ++i) {
+      step_scopes->push_back(std::make_shared<Scope>(scope));
+    }
+  }
+}
+
+void RecurrentOp::CreateStepNet(ScopePtr scope) const {
+  Variable* var = scope->CreateVariable(net_name_);
+  auto step_net = GetAttr<std::string>("step_net");
+  // get the step net proto from the string.
+  // PADDLE_ENFORCE(
+  //   google::protobuf::TextFormat::ParseFromString(step_net,
+  //   &step_net_desc_));
+  // var->Reset<PlainNet>(new PlainNet(step_net_desc_));
+  // this is a fake net, it will be rewrite after the network has been merged.
+  NetDesc desc;
+  desc.name_ = "rnn_step_net";
+  var->Reset<PlainNet>(new PlainNet(desc));
+  // TODO add op descs
+}
+
+void RecurrentOp::SegmentInputs(ScopePtr scope) const {
+  PADDLE_ENFORCE(!inlinks_.empty(), "no real inputs are provided.");
+  Variable* scopes_var = scope->GetVariable(step_scopes_name_);
+  auto& step_scopes = *scopes_var->GetMutable<std::vector<Scope*>>();
+
+  auto dims = Input(scope, 0)->GetMutable<Tensor>()->dims();
+  int seq_len = dims[0];
+  int batch_size = dims[1];
+  for (size_t i = 0; i < inlinks_.size(); i++) {
+    auto input_dims = Input(scope, i)->GetMutable<Tensor>()->dims();
+    int input_dim = input_dims[2];
+    int length = batch_size * input_dim;
+    const float* scope_input =
+        Input(scope, i)->GetMutable<Tensor>()->data<float>();
+    for (int j = 0; j < seq_len; j++) {
+      Variable* input_var = step_scopes[j]->CreateVariable(inlinks_[i]);
+      Tensor* step_input_tensor = input_var->GetMutable<Tensor>();
+      float* step_input = step_input_tensor->mutable_data<float>(
+          make_ddim({batch_size, input_dim}), platform::CPUPlace());
+      std::memcpy(step_input, scope_input + j * length, length);
+    }
+  }
+}
+
+void RecurrentOp::ConcatOutputs(ScopePtr scope) const {
+  Variable* scopes_var = scope->GetVariable(step_scopes_name_);
+  auto& step_scopes = *scopes_var->GetMutable<std::vector<Scope*>>();
```

Comment:
`std::vector<Scope*>>` -> `std::vector<ScopePtr>` ?

scopes should manage `Scope`'s memory